### ISSUE TITLE
Song's cover art is not showing up

### ISSUE THREAD

**Issue by NamanOG** [PKBeam/AMWin-RP#47]
Earlier the Cover art used to appear on the song but now it's just Apple Music logo
Tried multiple songs of different artists but cover art is not showing.
![No Cover art showing up](https://github.com/PKBeam/AMWin-RP/assets/100674016/724fef57-c97b-4f26-9245-dd87b6028317)
![image](https://github.com/PKBeam/AMWin-RP/assets/100674016/04d98887-8c63-43ae-8011-9cfb5c4eb246)

**Comment by PKBeam** [comment]
If you had a Last.FM account connected, this probably will be fixed with v1.2.5 (364d4aa4e68a6c5f577431d490e591c8a5410055).

**Comment by kellwinr** [comment]
> If you had a Last.FM account connected, this probably will be fixed with v1.2.5 ([364d4aa](https://github.com/PKBeam/AMWin-RP/commit/364d4aa4e68a6c5f577431d490e591c8a5410055)).

Hi, I'm using v1.2.5 but the issue is still there.
![image](https://github.com/PKBeam/AMWin-RP/assets/96328543/f0b5a3b9-ec6f-40ea-8d21-3750a6a3a13e)

**Comment by kiberOgur4ik** [comment]
> > If you had a Last.FM account connected, this probably will be fixed with v1.2.5 ([364d4aa](https://github.com/PKBeam/AMWin-RP/commit/364d4aa4e68a6c5f577431d490e591c8a5410055)).
> 
> Hi, I'm using v1.2.5 but the issue is still there. ![image](https://user-images.githubusercontent.com/96328543/270297888-f0b5a3b9-ec6f-40ea-8d21-3750a6a3a13e.png)

It worked after turning on last.fm.  The app takes cover arts from there now, but it works slowly and sometimes shows incorrect cover arts.

**Comment by kellwinr** [comment]
> > > If you had a Last.FM account connected, this probably will be fixed with v1.2.5 ([364d4aa](https://github.com/PKBeam/AMWin-RP/commit/364d4aa4e68a6c5f577431d490e591c8a5410055)).
> > 
> > 
> > Hi, I'm using v1.2.5 but the issue is still there. ![image](https://user-images.githubusercontent.com/96328543/270297888-f0b5a3b9-ec6f-40ea-8d21-3750a6a3a13e.png)
> 
> It worked after turning on last.fm. The app takes cover arts from there now, but it works slowly and sometimes shows incorrect cover arts.

My last.fm is connected and working, but not a single time an album art shows up in discord RP

**Comment by tyleraluko** [comment]
Wanted to jump in and add that I don't have Last.fm connected and also disabled the setting in the RP app; still no album art here. :(

**Comment by kellwinr** [comment]
update: the album cover art showed up after restarting AMWin and AM
![goh](https://github.com/PKBeam/AMWin-RP/assets/96328543/decfb963-bd0a-4828-bc2f-dd491eb057ec)

**Comment by Da532** [comment]
Has anyone discovered a solution to this or is it an unavoidable bug?

I've tried restarting both the RPC client and AM. No dice.

**Comment by Luke-Ert** [comment]
Same issue here, any fix for those not using last fm scrobbles?

**Comment by PKBeam** [comment]
Cover art searching is imprecise because we only have access to song and artist name strings. There's no unique song identifier exposed to us from the Apple Music app's UI; we have to infer the song from string structures, which is inherently problematic.

Without a Last.FM key, the app will GET the Apple Music web search page and scrape the HTML for results. This is slow and unreliable because we have to download and parse an entire webpage instead of just calling an API. Apple Music has an API but it requires a paid Apple developer account. 

With a Last.FM key, we can call the Last.FM API to search for album art. This performs better but still has the above problems with precision. 

In either case we have to check that the search results have the correct song and artist. But again we need to work with string structures and there are lots of small formatting differences between the Apple Music Windows app, the Apple Music web app, and the Last.FM API. It's really hard to find all these differences so there will probably be issues with multiple artists, symbols or non-ASCII characters.

There are possible improvements that can be made:
- caching album art URLs to memory/disk,
- invalidating pending album art requests on a song change,
- being more judicious about how often HTTP requests are sent,
- switching to a third party API that doesn't require credentials like Last.FM.  

If anyone wants to try and implement these I'm happy to accept a PR.

TL;DR because we can't directly get song info we have to do it in roundabout ways. This introduces tons of uncertainties and possible sources of error.
